### PR TITLE
spinner: add custom loading animation

### PR DESCRIPTION
**Pull request by vince-fugnitto**
#### What it does

Fixes: #72 

The following pull-request introduces a `preloadTemplate` to the application to provide a custom loading animation (splash screen) which is different than upstream, and can be used as an example for others:

<details>

<summary>Videos</summary>

<br />

https://user-images.githubusercontent.com/40359487/114600920-e6edbc00-9c62-11eb-899a-6dc0658241e6.mp4

<br />

https://user-images.githubusercontent.com/40359487/114600943-ec4b0680-9c62-11eb-9916-d62cc29f694b.mp4

</details>

I am open to suggestions on what we might want the `svg` to be :+1: 

#### How to test

Start or package the application, the new loading screen should take effect.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

**Comment by marcdumais-work** [comment]
How about adding "Blueprint" as part of the spinner svg? Or make something re-using the Blueprint logo?

**Comment by vince-fugnitto** [comment]
> How about adding "Blueprint" as part of the spinner svg? Or make something re-using the Blueprint logo?

I tried adding blueprint to the spinner itself (locally), in theory we can have anything :)


https://user-images.githubusercontent.com/40359487/114899646-217d6300-9de1-11eb-8ed2-3400dc40801b.mp4

**Comment by brianking** [comment]
Idea: An animated gif of the blueprint logo, with lines being drawn on it.

**Comment by vince-fugnitto** [comment]
> Idea: An animated gif of the blueprint logo, with lines being drawn on it.

@brianking I like the idea. but unfortunately I do not have the experience or knowledge on how to make more complex animations. Perhaps there is someone at the foundation who can make such animations?

**Comment by brianking** [comment]
@vince-fugnitto Sure, I will followup. 

What are the size specifications?

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto Sure, I will followup.
> What are the size specifications?

Given it is an animated `svg` like I have proposed we can resize it without issues and it would be responsive.
Else, we might want to keep it relatively modest (I'm not sure about the specifics).

**Comment by marcdumais-work** [comment]
> an animated `svg` like I have proposed we can resize it without issues

Correct. So the question might be more like: what's the lower range in size, we might use, so that the details are visible? I don't personally think we necessarily need it to be postage-stamp small.

**Comment by croundy** [comment]
Please see if any of these resemble the direction you are thinking
[Theia_logo_and_icon_animations_2 (1).zip] (https://github.com/eclipse-theia/theia-blueprint/files/6394189/Theia_logo_and_icon_animations_2.1.zip)

**Comment by marcdumais-work** [comment]
> Please see if any of these resemble the direction you are thinking
> [Theia_logo_and_icon_animations_2 (1).zip] (https://github.com/eclipse-theia/theia-blueprint/files/6394189/Theia_logo_and_icon_animations_2.1.zip)

@croundy Nice start. I suggest a quicker animation. This will be displayed while Blueprint loads, which can be relatively fast. Maybe a couple of seconds, typically. So, most users may never see the best parts of a ~8s animation :)

Maybe we try to get to this point within a second: 
image

**Comment by koegel** [comment]
@croundy I like option B. I agree with @marcdumais-work that it should be much faster though. Also the transition between the separated parts and the final logo could be smoother if possible.

**Comment by croundy** [comment]
Here's the update from Mehrnaz, as well as a new option she put together. [Theia_Logo_Motion (1).zip](https://github.com/eclipse-theia/theia-blueprint/files/6429838/Theia_Logo_Motion.1.zip)

**Comment by vince-fugnitto** [comment]
> Here's the update from Mehrnaz, as well as a new option she put together. [Theia_Logo_Motion (1).zip](https://github.com/eclipse-theia/theia-blueprint/files/6429838/Theia_Logo_Motion.1.zip)

@croundy I don't think the filetypes provided will work (`mp4`), it should be an animated svg.

**Comment by croundy** [comment]
@vince-fugnitto we will have Mehrnaz create whatever file type is needed. For now, just looking for feedback on the actual animation and if either of these options are in line with what you guys are thinking

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto we will have Mehrnaz create whatever file type is needed. For now, just looking for feedback on the actual animation and if either of these options are in line with what you guys are thinking

Thank you for clarifying, I wouldn't want Mehrnaz to perform unnecessary work 👍 

I don't have the strongest opinion on the latest animation, but I do not get the sense of loading or progression.
We must remember that applications can either start quickly or take some time. In both cases we should have an animation which displays the logo, and can be properly looped. Unfortunately I do not believe this animation can be looped cleanly for longer startup times.

**Comment by koegel** [comment]
https://user-images.githubusercontent.com/1044916/117508658-4355aa00-af89-11eb-89f1-32ca87a7d631.mp4

**Comment by koegel** [comment]
https://user-images.githubusercontent.com/1044916/117508721-59fc0100-af89-11eb-8119-5c1d1ba81ee3.mp4

**Comment by croundy** [comment]
> @koegel I want warp-jump blueprint now :)
> (new option 1)

Ya, I kind of like warp jump too, but I will leave it to you guys to decide

**Comment by croundy** [comment]
See what you think of these options: 

https://user-images.githubusercontent.com/72274403/117978644-5650dd00-b2ef-11eb-96bd-cfb8d5e9333f.mp4

https://user-images.githubusercontent.com/72274403/117978681-6072db80-b2ef-11eb-9838-1fcb71b7de78.mp4

https://user-images.githubusercontent.com/72274403/117978710-67015300-b2ef-11eb-9fb4-25402a63ecf4.mp4

**Comment by brianking** [comment]
I like the 3rd one the best.

**Comment by koegel** [comment]
> I like the 3rd one the best.

So do I!

**Comment by marcdumais-work** [comment]
Personally I'd like to have the name "Theia Blueprint" present, from the beginning until the end. I'll go with what the majority likes.

**Comment by jfaltermeier** [comment]
@vince-fugnitto This is part of the Beta 1 milestone. We plan to move from Alpha to Beta end of June. Shall we move this ticket?

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto This is part of the Beta 1 milestone. We plan to move from Alpha to Beta end of June. Shall we move this ticket?

@jfaltermeier if you are referring to updating the milestone for the pull-request then please go ahead! :)

**Comment by brianking** [comment]
We'll need someone to make a final decision on this one.

**Comment by marcdumais-work** [comment]
> We'll need someone to make a final decision on this one.

Let's go with `D.mp4` above, AKA `"third one"`. @croundy I think we'll need the animation is `SVG` format. If you can provide it, Vince can update this PR.

**Comment by brianking** [comment]
I have requested an SVG version.

**Comment by vince-fugnitto** [comment]
> I have requested an SVG version. Would an animated GIF work as well?

@brianking an animated gif can possibly work but it would not be of high quality, clear, or as responsive as an `svg` would.

**Comment by marcdumais-work** [comment]
@brianking I think this fell between the cracks. Can you see if it's still possible to get an `svg` version of the selected animation?

**Comment by brianking** [comment]
Here is the animated GIF version, is that acceptable?

D

The Theia Blueprint logo is a raster image (with pattern) and not a vector file, converting a raster pattern file to SVG is not good practice, as patterns needs lots of attributes and Id to define on both html5, and css files. Some information will be lost in the process.

**Comment by marcdumais-work** [comment]
Thanks @brianking - let's try with this.

**Comment by marcdumais-work** [comment]
@vince-fugnitto I briefly tried to update this PR to use the provided GIF but was unsuccessful. Would you have a try at it?

**Comment by vince-fugnitto** [comment]
@marcdumais-work the pull-request is rebased so it should now work :+1: 
@brianking I do not believe the https://github.com/eclipse-theia/theia-blueprint/pull/73#issuecomment-982858180 would work, for instance it has a white background that would not look well on dark themes, and would not scale well like an svg.

**Comment by brianking** [comment]
We have made another gif animation with lighter pattern's color that works good both on dark and light platforms. As well, we made an svg animation from the theia icon as it is a vector image and can be converted to svg well.  ( As discussed the logo is a raster image  since it has a pattern. Raster images when converting to svgs created several faulty codes) The new svg animation have been saved on a root folder that attached here. Please review and open the index.html in a browser. let us know your thoughts.

final Theia animation

[theiaSvg.zip](https://github.com/eclipse-theia/theia-blueprint/files/7719015/theiaSvg.zip)

**Comment by vince-fugnitto** [comment]
@marcdumais-work I updated the pull-request to use the latest animation provided, let me know what you think :)

**Comment by marcdumais-work** [comment]
Thanks @vince-fugnitto !

I think I like the simplicity. Is there a way to make the movement smoother? Here's how it looks on my machine:

Peek 2021-12-15 10-29

**Comment by marcdumais-work** [comment]
I discussed with Vince offline. The jerkiness of the animation is a known issue (1). Thinking of it, it's displayed at startup, when the browser app is very busy setting everything up and loading the workspace and so on. As noted by Thomas, the JSON-RPC communication currently suffers a pretty big encoding/decoding overhead, the price of which is probably very high right at startup. 

TL;DR: in the short term we have to live with a startup animation slowdowns. We can try for an animation, such that the effect of slowdowns will be less visually prominent, probably by having the logo stay static and limiting the animation to secondary element(s). Vince might have an additional idea or two. 

(1): https://github.com/eclipse-theia/theia/discussions/9287

**Comment by vince-fugnitto** [comment]
@marcdumais-work I updated the pull-request to go a different route which should not have the same low frame rates on startup (using [`@keyframes`](https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes)):

https://user-images.githubusercontent.com/40359487/146226499-c4378ea8-846f-4279-a3da-387d7a860f54.mp4

**Comment by marcdumais-work** [comment]
Here I composed a busy workbench, which means lots to restore when I reload (`CTRL-R`), so we can observe the latest version of the animation. I quite like it, it's pulsating like it's alive, breathing. 

Peek 2021-12-15 14-32

@vince-fugnitto should we try bigger?

**Comment by vince-fugnitto** [comment]
The pull-request should be ready for anyone interested in reviewing :)

**Review by jfaltermeier — APPROVED** [review]
I've testing on Linux. Looks very nice. I could not observe any lags during the animation.
